source/server/server.h

```diff
@@ -200,6 +202,10 @@ class ServerFactoryContextImpl : public Configuration::ServerFactoryContext,
   envoy::config::bootstrap::v3::Bootstrap& bootstrap() override { return server_.bootstrap(); }
   OverloadManager& overloadManager() override { return server_.overloadManager(); }
   bool healthCheckFailed() const override { return server_.healthCheckFailed(); }
+  Configuration::DownstreamHTTPFilterConfigProviderManagerSharedPtr
+  downstreamHttpFilterConfigProviderManager() override {
```

Comment:
I am not clear on why we need this in the server context where we could do ECDS for other filters without adding a server accessor.  If this is only  needed for the composite filter can the composiite filter factory create and own instead?

Reply:
right but why should it be in the server API?  Doesn't the listener factory have listener-specific APIs for this purpose?
/wait

Reply:
We started with ListenerFactory but moved to ServerContext because it has to span across listeners. See this [discussion](https://github.com/envoyproxy/envoy/pull/29289#discussion_r1387526023)

Reply:
Sorry for the delay.  I see the discussion there but I'm still confused why the composite filter needs  global state here where every other ECDS-aware filter does not.  Are the other ones buggy?  If so can we file a tracking bug?  If not I'm still not clea on why the composite filter can't get ECDS updates the same way every other filter does without more server-wide state
cc @adisuissa

Reply:
AFAICT previously the HTTP-filter-config-provider was a singleton in the HCM, so any HCM was able to access the same instance in order to dynamically fetch configs. Now the config-provider is a data-member of the ServerFactoryContext, and a getter allows access to it.

If the prior behavior is desired (i.e., only if the HCM is used then the singletons will be created), and limited "accessibility" (avoid being globally in the ServerFactoryContext), then I think an alternative could be to keep the [creation of a singleton](https://github.com/envoyproxy/envoy/blob/7c6a3822b33b326bc1b71ba482afa9927420af7e/source/extensions/filters/network/http_connection_manager/config.cc#L234-L236) in the HCM, and plumb the HTTP-filter-config-provider to the [HttpFilterActionContext](https://github.com/envoyproxy/envoy/blob/7c6a3822b33b326bc1b71ba482afa9927420af7e/source/common/http/matching/data_impl.h#L69C8-L69C31) object, which is accessible by the Composite-Filter.
Not sure if there are other tradeoffs, though.

Reply:
Initially I started with holding `DynamicConfigProvider` in `ActionFactory` that resulted in life cycle issues during destruction - stats scope is not cleaned up

```
[2023-08-30 03:27:51.950][1359028][critical][assert] [source/common/stats/thread_local_store.cc:49] assert failure: scopes_.empty().
[2023-08-30 03:27:51.950][1359028][critical][backtrace] [./source/server/backtrace.h:104] Caught Abort trap: 6, suspect faulting address 0x1a9348764
[2023-08-30 03:27:51.950][1359028][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2023-08-30 03:27:51.950][1359028][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.28.0-dev/test/DEBUG/BoringSSL
[2023-08-30 03:27:51.967][1359028][critical][backtrace] [./source/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x1061dfeb8]
[2023-08-30 03:27:51.967][1359028][critical][backtrace] [./source/server/backtrace.h:96] #1: _sigtramp [0x1a93aea24]
[2023-08-30 03:27:51.967][1359028][critical][backtrace] [./source/server/backtrace.h:96] #2: pthread_kill [0x1a937fc28]
[2023-08-30 03:27:51.967][1359028][critical][backtrace] [./source/server/backtrace.h:96] #3: abort [0x1a928dae8]
[2023-08-30 03:27:51.969][1359028][critical][backtrace] [./source/server/backtrace.h:96] #4: Envoy::Stats::ThreadLocalStoreImpl::~ThreadLocalStoreImpl() [0x102a0ec40]
[2023-08-30 03:27:51.971][1359028][critical][backtrace] [./source/server/backtrace.h:96] #5: Envoy::Stats::ThreadLocalStoreImpl::~ThreadLocalStoreImpl() [0x102a0fa88]
[2023-08-30 03:27:51.972][1359028][critical][backtrace] [./source/server/backtrace.h:96] #6: Envoy::IntegrationTestServerImpl::createAndRunEnvoyServer() [0x1016506c4]
[2023-08-30 03:27:51.973][1359028][critical][backtrace] [./source/server/backtrace.h:96] #7: Envoy::IntegrationTestServer::threadRoutine() [0x10164f990]
[2023-08-30 03:27:51.975][1359028][critical][backtrace] [./source/server/backtrace.h:96] #8: Envoy::IntegrationTestServer::start()::$_2::operator()() [0x10165e340]
[2023-08-30 03:27:51.976][1359028][critical][backtrace] [./source/server/backtrace.h:96] #9: _ZNSt3__1L8__invokeIRZN5Envoy21IntegrationTestServer5startENS1_7Network7Address9IpVersionENS_8functionIFvvEEENS_8optionalIyEEbNS9_INS_17reference_wrapperINS1_13ProcessObjectEEEEENS1_6Server21FieldValidationConfigEjNS_6chrono8durationIxNS_5ratioILl1ELl1EEEEENSF_13DrainStrategyENS_10shared_ptrINS1_6Buffer16WatermarkFactoryEEEbE3$_2JEEEDTclscT_fp_spscT0_fp0_EEOST_DpOSU_ [0x10165e250]
[2023-08-30 03:27:51.977][1359028][critical][backtrace] [./source/server/backtrace.h:96] #10: std::__1::__invoke_void_return_wrapper<>::__call<>() [0x10165e208]
[2023-08-30 03:27:51.978][1359028][critical][backtrace] [./source/server/backtrace.h:96] #11: std::__1::__function::__alloc_func<>::operator()() [0x10165e1e0]
[2023-08-30 03:27:51.980][1359028][critical][backtrace] [./source/server/backtrace.h:96] #12: std::__1::__function::__func<>::operator()() [0x10165cca4]
[2023-08-30 03:27:51.981][1359028][critical][backtrace] [./source/server/backtrace.h:96] #13: std::__1::__function::__value_func<>::operator()() [0x1011da004]
[2023-08-30 03:27:51.982][1359028][critical][backtrace] [./source/server/backtrace.h:96] #14: std::__1::function<>::operator()() [0x1011d9d2c]
[2023-08-30 03:27:51.984][1359028][critical][backtrace] [./source/server/backtrace.h:96] #15: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::operator()() [0x10729eef8]
[2023-08-30 03:27:51.985][1359028][critical][backtrace] [./source/server/backtrace.h:96] #16: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::__invoke() [0x10729eecc]
[2023-08-30 03:27:51.985][1359028][critical][backtrace] [./source/server/backtrace.h:96] #17: _pthread_start [0x1a937ffa8]
[2023-08-30 03:27:51.985][1359028][critical][backtrace] [./source/server/backtrace.h:96] #18: thread_start [0x1a937ada0]
```
So me and @kyessenov discussed and decided to move this to serverfactory context and let HCM also use the same. So all Http filter related ECDS would use this.

I do not know if keeping it as singleton and wiring up to [HttpFilterActionContext](https://github.com/envoyproxy/envoy/blob/7c6a3822b33b326bc1b71ba482afa9927420af7e/source/common/http/matching/data_impl.h#L69C8-L69C31) would provide any additional benefits than having it in ServerContext. 

But if that is better - I am certainly up for changing it.

Reply:
yeah after some conversation with Adi I'm ok with this as is.